### PR TITLE
Implementation of backup and restore APIs.

### PR DESCRIPTION
**Pull request by branupama**
This includes implementation of some of BackupRestore class APIs and triggering backup and restore workflow in different scenario based on the system type.

### CODE REVIEW

File: include/backup_restore.hpp [hunk 1]
```diff
@@ -55,6 +55,14 @@ class BackupRestore
      */
     std::tuple<types::VPDMapVariant, types::VPDMapVariant> backupAndRestore();
 
+    /**
+     * @brief An API to update whether backup and restore functionality is being
```
Comment (SunnySrivastava1984) [inline include/backup_restore.hpp:59]: brief needs to be wrapped with in a line.

File: include/backup_restore.hpp [hunk 2]
```diff
@@ -55,6 +55,14 @@ class BackupRestore
...
+     * called or not.
+     *
+     * @param[in] i_isCalled - Status to set.
+     */
+    void setIsBackupAndRestoreCalled(bool i_isCalled);
```
Comment (SunnySrivastava1984) [inline include/backup_restore.hpp:64]: > setIsBackupAndRestoreCalled

setBackupAndRestoreStatus?

File: include/backup_restore.hpp [hunk 3]
```diff
@@ -69,7 +77,13 @@ class BackupRestore
...
                                 const std::string& i_dstPath);
 
+    // Parsed system JSON config file.
+    nlohmann::json m_sysCfgJsonObj;
+
     // Parsed backup and restore's JSON config file.
-    nlohmann::json m_backupRestoreParsedJson{};
+    nlohmann::json m_backupRestoreCfgJsonObj{};
+
+    // Indicates whether backup and restore is being called already or not.
```
Comment (SunnySrivastava1984) [inline include/backup_restore.hpp:86]: > // Indicates whether backup and restore is being called already or not. -> grammar

// Indicates if backup and restore has been performed.

File: include/backup_restore.hpp [hunk 4]
```diff
@@ -69,7 +77,13 @@ class BackupRestore
...
+    bool m_isBackupAndRestoreCalled = false;
```
Comment (SunnySrivastava1984) [inline include/backup_restore.hpp:87]: > m_isBackupAndRestoreCalled

m_backupAndRestoreDone

File: include/worker.hpp [hunk 5]
```diff
@@ -356,5 +362,8 @@ class Worker
 
     // Path to config JSON if applicable.
     std::string& m_configJsonPath;
+
+    // Holds backup restore object
+    BackupRestore* m_backupRestoreObjPtr = nullptr;
```
Comment (SunnySrivastava1984) [inline include/worker.hpp:367]: Why class level pointer required?

File: src/manager.cpp [hunk 6]
```diff
@@ -116,6 +116,7 @@ void Manager::SetTimerToDetectSVPDOnDbus()
             m_worker->collectFrusFromJson();
             m_interface->set_property("CollectionStatus",
                                       std::string("Completed"));
+            m_worker->backupAndRestore();
```
Comment (SunnySrivastava1984) [inline src/manager.cpp:119]: Did not understand this. Why manager is calling via worker?

File: src/worker.cpp [hunk 7]
```diff
@@ -1193,4 +1214,14 @@ void Worker::collectFrusFromJson()
         }
     }
 }
+
+std::tuple<types::VPDMapVariant, types::VPDMapVariant>
```
Comment (SunnySrivastava1984) [inline src/worker.cpp:1218]: Why is this API required? Why worker is even exposing an API to do back up and restore?

File: src/backup_restore.cpp [hunk 8]
```diff
@@ -23,4 +162,9 @@ void BackupRestore::backupAndRestoreIpzVpd(types::IPZVpdMap& io_srcVpdMap,
     (void)i_srcPath;
     (void)i_dstPath;
 }
+
+void BackupRestore::setIsBackupAndRestoreCalled(bool i_isCalled)
```
Comment (SunnySrivastava1984) [inline src/backup_restore.cpp:166]: isCalled does not seem like a proper name.

File: include/worker.hpp [hunk 9]
```diff
@@ -100,6 +101,11 @@ class Worker
                       types::ObjectMap& objectInterfaceMap,
                       const std::string& vpdFilePath);
 
+    /**
+     * @brief An API to backup and restore VPD.
+     */
+    std::tuple<types::VPDMapVariant, types::VPDMapVariant> backupAndRestore();
```
Comment (SunnySrivastava1984) [inline include/worker.hpp:107]: Don't see a need of this public API.

File: src/worker.cpp [hunk 10]
```diff
@@ -486,11 +486,32 @@ void Worker::setDeviceTreeAndJson()
 
     if (fitConfigVal.find(devTreeFromJson) != std::string::npos)
     { // fitconfig is updated and correct JSON is set.
-
         if (isSystemVPDOnDBus())
         {
-            // TODO
-            //  Restore system VPD logic should initiate from here.
+            m_backupRestoreObjPtr = new BackupRestore(m_parsedJson);
```
Comment (SunnySrivastava1984) [inline src/worker.cpp:491]: Why is this needed and where it is getting freed.

File: src/worker.cpp [hunk 11]
```diff
@@ -486,11 +486,32 @@ void Worker::setDeviceTreeAndJson()
...
+            if (auto l_sysVpd = std::get_if<types::IPZVpdMap>(&parsedVpdMap))
+            {
+                const std::string& l_sysImKwdValue = getIMValue(*l_sysVpd);
+
+                /* System's list, for which backup and restore should be called
+                 * before publishing VPD on DBus.*/
+                const std::vector<std::string> l_systemTypes = {
```
Comment (SunnySrivastava1984) [inline src/worker.cpp:498]: Why the check is after instantiating the class?

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: include/backup_restore.hpp (hunk 1 above)
Comment (branupama) [reply include/backup_restore.hpp:59]: Done

File: include/backup_restore.hpp (hunk 2 above)
Comment (branupama) [reply include/backup_restore.hpp:64]: Done

File: include/backup_restore.hpp (hunk 3 above)
Comment (branupama) [reply include/backup_restore.hpp:86]: Done

File: include/backup_restore.hpp (hunk 4 above)
Comment (branupama) [reply include/backup_restore.hpp:87]: Done

File: include/worker.hpp (hunk 9 above)
Comment (branupama) [reply include/worker.hpp:107]: Removed this API, this PR is only including BackupRestore API's implementation.

File: include/worker.hpp (hunk 5 above)
Comment (branupama) [reply include/worker.hpp:367]: Removed

File: src/backup_restore.cpp (hunk 8 above)
Comment (branupama) [reply src/backup_restore.cpp:166]: Updated

File: src/manager.cpp (hunk 6 above)
Comment (branupama) [reply src/manager.cpp:119]: Removed this statement, only including BackupRestore API's implementation in this PR.

File: src/worker.cpp (hunk 10 above)
Comment (branupama) [reply src/worker.cpp:491]: only including BackupRestore API's implementation in this PR.

File: src/worker.cpp (hunk 11 above)
Comment (branupama) [reply src/worker.cpp:498]: only including BackupRestore API's implementation in this PR.

File: src/worker.cpp (hunk 7 above)
Comment (branupama) [reply src/worker.cpp:1218]: only including BackupRestore API's implementation in this PR.

File: src/backup_restore.cpp [hunk 12]
```diff
@@ -1,16 +1,155 @@
...
+                m_backupRestoreCfgJsonObj["source"].value("hardwarePath", "");
+            !l_srcHwFilePath.empty() &&
+            std::filesystem::exists(l_srcHwFilePath))
+        {
+            std::shared_ptr<Parser> l_vpdParser =
+                std::make_shared<Parser>(l_srcHwFilePath, m_sysCfgJsonObj);
+            l_srcVpdMap = l_vpdParser->parse();
+            l_srcVpdPath = l_srcHwFilePath;
+        }
+        else if (const std::string& l_srcInventoryObjPath =
```
Comment (SunnySrivastava1984) [inline src/backup_restore.cpp:68]: > l_srcInventoryObjPath

We can get rid of this variable and a string copy after that by using l_srcVpdPath directly?

File: src/backup_restore.cpp [hunk 13]
```diff
@@ -1,16 +1,155 @@
...
+            !m_backupRestoreCfgJsonObj.contains("backupMap"))
+        {
+            logging::logMessage(
+                "Backup restore config JSON is missing necessary tag(s), can't initiate backup and restore.");
+            return l_emptyVariantPair;
+        }
+
+        std::string l_srcVpdPath;
+        types::VPDMapVariant l_srcVpdMap;
+        if (const std::string& l_srcHwFilePath =
```
Comment (SunnySrivastava1984) [inline src/backup_restore.cpp:58]: > l_srcHwFilePath

We can get rid of this variable and a string copy after that by using l_srcVpdPath directly?

File: src/backup_restore.cpp [hunk 14]
```diff
@@ -1,16 +1,155 @@
...
+        else
+        {
+            logging::logMessage(
+                "Couldn't extract source path, can't initiate backup and restore.");
+            return l_emptyVariantPair;
+        }
+
+        std::string l_dstVpdPath;
+        types::VPDMapVariant l_dstVpdMap;
+        if (const std::string& l_dstHwFilePath =
```
Comment (SunnySrivastava1984) [inline src/backup_restore.cpp:84]: Same in the case of destination?

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: src/backup_restore.cpp (hunk 14 above)
Comment (branupama) [reply src/backup_restore.cpp]: Updated.

File: src/backup_restore.cpp (hunk 13 above)
Comment (branupama) [reply src/backup_restore.cpp]: Done

File: src/backup_restore.cpp (hunk 12 above)
Comment (branupama) [reply src/backup_restore.cpp]: Done

File: src/backup_restore.cpp [hunk 15]
```diff
@@ -1,16 +1,155 @@
...
+#include "utils.hpp"
+
 namespace vpd
 {
-BackupRestore::BackupRestore(const nlohmann::json& i_parsedSystemJson)
+BackupRestore::BackupRestore(const nlohmann::json& i_sysCfgJsonObj) :
+    m_sysCfgJsonObj(i_sysCfgJsonObj)
 {
-    (void)i_parsedSystemJson;
+    try
```
[the remark this answers is not in the document]
Comment (branupama) [reply src/backup_restore.cpp]: Removed try & catch block, if any exception will be thrown directly to Worker.
So for any exception caught in backup and restore parsing config file will stop the application.

File: include/backup_restore.hpp [hunk 16]
```diff
@@ -25,10 +25,10 @@ class BackupRestore
     /**
      * @brief Constructor.
      *
-     * @param[in] i_parsedSystemJson - Parsed system config JSON.
+     * @param[in] i_sysCfgJsonObj - Parsed system config JSON.
```
Comment (SunnySrivastava1984) [inline include/backup_restore.hpp:28]: Constructor is throwing exception. That needs to be updated now.

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]
Commit message need to be updated correctly.

File: include/backup_restore.hpp (hunk 16 above)
Comment (branupama) [reply include/backup_restore.hpp:28]: done.

**Comment by branupama** [comment]
> Commit message need to be updated correctly.

Done

File: include/backup_restore.hpp (hunk 16 above)
Comment (SunnySrivastava1984) [reply include/backup_restore.hpp:28]: The correct wat is to add "@throw" which is a valid doxygen tag.
Please use that.

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: include/backup_restore.hpp (hunk 16 above)
Comment (branupama) [reply include/backup_restore.hpp:28]: Updated

**Review by SunnySrivastava1984 — APPROVED** [review]

File: include/backup_restore.hpp [hunk 17]
```diff
@@ -69,7 +78,15 @@ class BackupRestore
...
 
+    // Parsed system JSON config file.
+    nlohmann::json m_sysCfgJsonObj{};
+
     // Parsed backup and restore's JSON config file.
-    nlohmann::json m_backupRestoreParsedJson{};
+    nlohmann::json m_backupRestoreCfgJsonObj{};
+
+    // Indicates if backup and restore has been performed.
+    static bool m_backupAndRestoreDone;
```
Comment (jinuthomas) [inline include/backup_restore.hpp:88]: m_isBackupAndRestoreCalled  , I felt better here

**Review by jinuthomas — APPROVED** [review]
